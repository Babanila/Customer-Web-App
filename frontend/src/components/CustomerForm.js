import React from 'react'
import { cx, css } from 'emotion'
import { useHistory } from 'react-router-dom'
import FormComponent from './FormComponent'
import SingleButton from './SingleButton'
import BackButton from './BackButton'
import { initialState, baseUrl, dataPoster } from './Utils'

const customerFormDivStyle = css`
  margin-top: 7em;
  width: 400px;
  height: auto;
  display: flex;
  align-self: center;
  flex-direction: column;
  border: 1px solid #f54b3a;
  border-radius: 5px;
  z-index: 10;
`

const customerFormDivTitle = css`
  align-self: center;
  font-size: 2em;
  padding: 20px 0px;
`

const customerFormBtnGroup = css`
  margin-bottom: 20px;
  width: 65%;
  min-width: 260px;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-self: center;
  box-sizing: border-box;
  outline: none;
  background-color: #ffffff;
  color: #000000;
  font-size: 1.2em;
`

const addBtnStyle = css`
  padding: 10px;
  font-size: 1em;
  border: 2px solid #009900;
  border-radius: 5px;
  outline: none;
  color: #009900;
  background-color: #ffffff;
  &:hover {
    cursor: pointer;
    color: #ffffff;
    background-color: #009900;
  }
`

const resetBtnStyle = css`
  padding: 10px;
  font-size: 1em;
  outline: none;
  border: 2px solid #ff0000;
  border-radius: 5px;
  color: #ff0000;
  background-color: #ffffff;
  &:hover {
    cursor: pointer;
    color: #ffffff;
    background-color: #ff0000;
  }
`

function CustomerForm() {
  const history = useHistory()
  const [formState, setFormState] = React.useState(initialState)

  const handleInputChange = ({ target }) => {
    const { name, value } = target
    setFormState((state) => ({
      ...state,
      name: {
        ...state.name,
        first: name === 'first' ? value : state.name.first,
        last: name === 'last' ? value : state.name.last
      },
      [name]: value
    }))
  }

  const handleSubmit = async (e) => {
    e.preventDefault()
    try {
      const { data } = await dataPoster(`${baseUrl}/new_customer`, 'post', formState)
      alert(data)
    } catch (error) {
      console.log(error)
      error.message.includes(404) ? alert('Error during creation !!!') : alert(error.message)
    }
  }

  const handleReset = (e) => {
    e.preventDefault()
    setFormState(initialState)
  }

  const handleBack = () => history.push(`/`)

  return (
    <div className={cx(customerFormDivStyle)}>
      <div className={cx(customerFormDivTitle)}>Customer Registration Form</div>
      <FormComponent
        formData={formState}
        handleSubmit={handleSubmit}
        handleInputChange={handleInputChange}
      />
      <div className={cx(customerFormBtnGroup)}>
        <SingleButton btnName="Register" btnClick={handleSubmit} btnStyles={addBtnStyle} />
        <SingleButton btnName="Reset" btnClick={handleReset} btnStyles={resetBtnStyle} />
        <BackButton handleBack={handleBack} />
      </div>
    </div>
  )
}

export default CustomerForm
