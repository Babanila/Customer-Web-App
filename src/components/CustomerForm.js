import React from 'react'
import { cx, css } from 'emotion'
import { useHistory } from 'react-router-dom'
import FormComponent from './FormComponent'
import SingleButton from './SingleButton'
import BackButton from './BackButton'
import { isoDateConverter } from './Utils'

const customerFormDivStyle = css`
  margin-top: 3em;
  width: 400px;
  height: auto;
  display: flex;
  align-self: center;
  flex-direction: column;
  border: 1px solid #f54b3a;
  border-radius: 5px;
`

const customerFormDivTitle = css`
  align-self: center;
  font-size: 2em;
  padding: 20px 0px;
`

const customerFormBtnGroup = css`
  margin-bottom: 20px;
  width: 65%;
  min-width: 260px;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-self: center;
  box-sizing: border-box;
  outline: none;
  background-color: #ffffff;
  color: #000000;
  font-size: 1.2em;
`

const addBtnStyle = css`
  padding: 10px;
  font-size: 1em;
  border: 2px solid #009900;
  border-radius: 5px;
  outline: none;
  color: #009900;
  background-color: #ffffff;
  &:hover {
    cursor: pointer;
    color: #ffffff;
    background-color: #009900;
  }
`

const resetBtnStyle = css`
  padding: 10px;
  font-size: 1em;
  outline: none;
  border: 2px solid #ff0000;
  border-radius: 5px;
  color: #ff0000;
  background-color: #ffffff;
  &:hover {
    cursor: pointer;
    color: #ffffff;
    background-color: #ff0000;
  }
`

const initialState = {
  customerID: '',
  name: {
    first: '',
    last: ''
  },
  birthday: '',
  gender: '',
  lastContact: '',
  customerLifetimeValue: 0
}

function CustomerForm() {
  const history = useHistory()
  const [formState, setFormState] = React.useState(initialState)

  const handleInputChange = ({ target }) => {
    const { name, value } = target
    let inputValue
    name === 'lastContact' ? (inputValue = isoDateConverter(value)) : (inputValue = value)
    setFormState((state) => ({
      ...state,
      name: {
        ...state.name,
        first: name === 'first' ? inputValue : state.name.first,
        last: name === 'last' ? inputValue : state.name.last
      },
      [name]: inputValue
    }))
  }

  const handleSubmit = (e) => {
    e.preventDefault()
    console.log('Submit button clicked')
  }

  const handleReset = (e) => {
    e.preventDefault()
    setFormState(initialState)
  }

  const handleBack = () => history.push(`/`)

  return (
    <div className={cx(customerFormDivStyle)}>
      <div className={cx(customerFormDivTitle)}>Customer Registration Form</div>
      <FormComponent
        formData={formState}
        handleSubmit={handleSubmit}
        handleInputChange={handleInputChange}
      />
      <div className={cx(customerFormBtnGroup)}>
        <SingleButton btnName="Register" btnClick={handleSubmit} btnStyles={addBtnStyle} />
        <SingleButton btnName="Reset" btnClick={handleReset} btnStyles={resetBtnStyle} />
        <BackButton handleBack={handleBack} />
      </div>
    </div>
  )
}

export default CustomerForm
