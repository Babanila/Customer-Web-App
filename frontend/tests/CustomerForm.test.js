import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import CustomerForm from '../src/components/CustomerForm'

describe('Check CustomerForm Component', () => {
  let sampleData, CustomerFormWrapper, match
  let actualDiv, actualDiv1, actualDiv2

  beforeEach(() => {
    sampleData = {
      customerID: 9,
      name: {
        first: 'John',
        last: 'Bull'
      },
      birthday: '1999-08-15',
      gender: 'w',
      lastContact: '2010-05-09T11:08:50.700Z',
      customerLifetimeValue: 2513
    }

    CustomerFormWrapper = mount(
      <MemoryRouter initialentries="{['/']}">
        <CustomerForm />
      </MemoryRouter>
    )

    actualDiv = CustomerFormWrapper.find('MemoryRouter')
    actualDiv1 = actualDiv.find('Router')
    actualDiv2 = actualDiv1.find('CustomerForm')
  })

  it('should renders component without crashing', () => {
    const tree = renderer.create(<CustomerForm />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render CustomerForm component without data', () => {
    const actualDiv3 = actualDiv2.find('div')

    expect(CustomerForm).toBeDefined()
    expect(actualDiv.length).toEqual(1)
    expect(actualDiv1.length).toEqual(1)
    expect(actualDiv2.length).toEqual(1)
    expect(actualDiv3.length).toEqual(13)
  })

  it('should render CustomerForm component data', () => {
    const setSelectedCustomer = React.useState
    jest.spyOn(React, 'useState').mockImplementationOnce(() => setSelectedCustomer(sampleData))

    const setLoading = React.useState
    jest.spyOn(React, 'useState').mockImplementationOnce(() => setLoading(false))

    const CustomerFormWrap = mount(
      <MemoryRouter initialentries="{['/']}">
        <CustomerForm />
      </MemoryRouter>
    )

    const actual1 = CustomerFormWrap.find('div')
    const actual2 = CustomerFormWrap.find('FormComponent')
    const actual3 = CustomerFormWrap.find('SingleButton')
    const actual4 = CustomerFormWrap.find('BackButton')
    const actual5 = actual2.children()

    expect(actual1.length).toEqual(13)
    expect(actual2.length).toEqual(1)
    expect(actual3.length).toEqual(3)
    expect(actual4.length).toEqual(1)
    expect(actual5.at(0).childAt(0).text()).toEqual('CustomerId:')
  })
})
