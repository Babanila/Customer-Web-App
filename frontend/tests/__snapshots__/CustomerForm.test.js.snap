// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check CustomerForm Component should renders component without crashing 1`] = `
<div
  className="css-111lz80"
>
  <div
    className="css-mqu2x4"
  >
    Customer Registration Form
  </div>
  <form
    className="css-cgyezj"
    onSubmit={[Function]}
  >
    <div
      className="css-ku1ga7"
    >
      <label>
        CustomerId:
        <input
          name="customerID"
          onChange={[Function]}
          placeholder="CustomerId"
          style={
            Object {
              "border": "none",
              "fontSize": "16px",
              "marginLeft": "0.5em",
              "outline": "none",
            }
          }
          type="text"
          value=""
        />
      </label>
    </div>
    <div
      className="css-ku1ga7"
    >
      <label>
        First Name:
        <input
          name="first"
          onChange={[Function]}
          placeholder="First Name"
          style={
            Object {
              "border": "none",
              "fontSize": "16px",
              "marginLeft": "0.5em",
              "outline": "none",
            }
          }
          type="text"
          value=""
        />
      </label>
    </div>
    <div
      className="css-ku1ga7"
    >
      <label>
        Last Name:
        <input
          name="last"
          onChange={[Function]}
          placeholder="Last Name"
          style={
            Object {
              "border": "none",
              "fontSize": "16px",
              "marginLeft": "0.5em",
              "outline": "none",
            }
          }
          type="text"
          value=""
        />
      </label>
    </div>
    <div
      className="css-ku1ga7"
    >
      <label>
        Gender:
      </label>
      <select
        name="gender"
        onChange={[Function]}
        style={
          Object {
            "border": "none",
            "fontSize": "16px",
            "marginLeft": "0.5em",
            "outline": "none",
          }
        }
        value=""
      >
        <option
          value=""
        >
           Choose one
        </option>
        <option
          value="m"
        >
          Male
        </option>
        <option
          value="w"
        >
          Female
        </option>
        <option
          value="o"
        >
          Other
        </option>
      </select>
    </div>
    <div
      className="css-ku1ga7"
    >
      <label>
        Birthday:
      </label>
      <div>
        <input
          name="birthday"
          onChange={[Function]}
          placeholder="Birthday"
          style={
            Object {
              "border": "none",
              "fontSize": "16px",
              "marginLeft": "0.5em",
              "outline": "none",
            }
          }
          type="date"
          value=""
        />
      </div>
    </div>
    <div
      className="css-ku1ga7"
    >
      <label>
        Last Contact:
      </label>
      <div>
        <input
          name="lastContact"
          onChange={[Function]}
          placeholder="Last Contact"
          style={
            Object {
              "border": "none",
              "fontSize": "16px",
              "marginLeft": "0.5em",
              "outline": "none",
            }
          }
          type="datetime-local"
          value=""
        />
      </div>
    </div>
    <div
      className="css-ku1ga7"
    >
      <label>
        Lifetime Value:
      </label>
      <div>
        <input
          name="customerLifetimeValue"
          onChange={[Function]}
          style={
            Object {
              "border": "none",
              "fontSize": "16px",
              "marginLeft": "0.5em",
              "outline": "none",
            }
          }
          type="text"
          value={0}
        />
      </div>
    </div>
  </form>
  <div
    className="css-kjw0s7"
  >
    <button
      className="css-gt9xug"
      onClick={[Function]}
    >
      Register
    </button>
    <button
      className="css-97sg0e"
      onClick={[Function]}
    >
      Reset
    </button>
    <button
      className="css-14mojuj"
      onClick={[Function]}
    >
      Back
    </button>
  </div>
</div>
`;
